### PR TITLE
First receiver version

### PR DESCRIPTION
**Pull request by pierDipi**
Signed-off-by: Pierangelo Di Pilato <pierangelodipilato@gmail.com>

### CODE REVIEW

**Comment by pierDipi** [comment]
cc @slinkydeveloper

File: .github/workflows/k8s_tests.yaml [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+          sudo apt-get install -y gettext-base
+          curl -Lo ./kind https://kind.sigs.k8s.io/dl/v0.8.1/kind-$(uname)-amd64
+          chmod +x ./kind
+          sudo mv ./kind /usr/local/bin/kind
+          kind version
+          curl -LO https://storage.googleapis.com/kubernetes-release/release/`curl -s https://storage.googleapis.com/kubernetes-release/release/stable.txt`/bin/linux/amd64/kubectl
+          chmod +x ./kubectl
+          sudo mv ./kubectl /usr/local/bin/kubectl
+          kubectl version --client
+      - name: Run k8s with KinD
```
Comment (slinkydeveloper) [inline .github/workflows/k8s_tests.yaml:19]: wooo kind in github actions!

File: .github/workflows/k8s_tests.yaml [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+jobs:
+  k8s_tests_data_plane:
+    runs-on: ubuntu-latest
+
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v2
+      - name: Install tools
+        run: |
+          sudo apt-get install -y gettext-base
```
Comment (slinkydeveloper) [inline .github/workflows/k8s_tests.yaml]: Can you put these commands in fancy scripts?

File: .github/workflows/k8s_tests.yaml [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+          chmod +x ./kind
+          sudo mv ./kind /usr/local/bin/kind
+          kind version
+          curl -LO https://storage.googleapis.com/kubernetes-release/release/`curl -s https://storage.googleapis.com/kubernetes-release/release/stable.txt`/bin/linux/amd64/kubectl
+          chmod +x ./kubectl
+          sudo mv ./kubectl /usr/local/bin/kubectl
+          kubectl version --client
+      - name: Run k8s with KinD
+        run: |
+          export KO_DOCKER_REPO=knative-kafka-broker
```
Comment (slinkydeveloper) [inline .github/workflows/k8s_tests.yaml]: same: fancy script

File: data-plane/build/build [hunk 4]
```diff
@@ -0,0 +1,70 @@
+#!/usr/bin/env bash
```
Comment (slinkydeveloper) [inline data-plane/build/build]: you forgot the extension of the file

File: data-plane/build/cleanup [hunk 5]
```diff
@@ -0,0 +1,5 @@
+#!/usr/bin/env bash
```
Comment (slinkydeveloper) [inline data-plane/build/cleanup]: file extension

File: data-plane/build/common [hunk 6]
```diff
@@ -0,0 +1,25 @@
+#!/usr/bin/env bash
```
Comment (slinkydeveloper) [inline data-plane/build/common]: file extension

File: data-plane/config/100-config-kafka-broker-receiver.yaml [hunk 7]
```diff
@@ -0,0 +1,38 @@
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: config-kafka-broker-receiver
+  namespace: knative-eventing
+data:
+  config-kafka-broker-producer: |
```
Comment (slinkydeveloper) [inline data-plane/config/100-config-kafka-broker-receiver.yaml]: I think this file should be named .properties

File: data-plane/receiver/src/main/java/dev/knative/eventingkafkabroker/receiver/CloudEventRequestToRecordMapper.java [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+            return;
+          }
+
+          // TODO(pierDipi) set the correct producer record key
+          final var record = KafkaProducerRecord.create(topic.get(), "", result.result());
+          recordConsumer.accept(Optional.of(record));
+        });
+  }
+
+  static Optional<String> topic(final String path) {
```
Comment (slinkydeveloper) [inline data-plane/receiver/src/main/java/dev/knative/eventingkafkabroker/receiver/CloudEventRequestToRecordMapper.java]: I'm not a fan of `Optional`s on the hot path honestly, JVM could not apply escape analysis on them, causing allocations

Comment (slinkydeveloper) [reply data-plane/receiver/src/main/java/dev/knative/eventingkafkabroker/receiver/CloudEventRequestToRecordMapper.java]: When you need to return null _on the hot path_, document it (even for private / package private methods)

File: data-plane/receiver/src/main/java/dev/knative/eventingkafkabroker/receiver/Main.java [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+
+  private static final Logger logger = LoggerFactory.getLogger(Main.class);
+
+  /**
+   * Start receiver.
+   *
+   * @param args command line arguments.
+   */
+  public static void main(final String[] args) {
+    final var env = new Env(System::getenv);
```
Comment (slinkydeveloper) [inline data-plane/receiver/src/main/java/dev/knative/eventingkafkabroker/receiver/Main.java]: If you need more complex configuration, keep in mind this: https://vertx.io/docs/vertx-config/java/#_using_the_config_retriever

It allows to retrieve configurations from system props, envs and config files transparently

File: data-plane/receiver/src/main/java/dev/knative/eventingkafkabroker/receiver/CloudEventRequestToRecordMapper.java [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+public class CloudEventRequestToRecordMapper implements RequestToRecordMapper<String, CloudEvent> {
+
+  static final int PATH_TOKEN_NUMBER = 2;
+  static final String PATH_DELIMITER = "/";
+  static final String TOPIC_DELIMITER = "-";
+
+  @Override
+  public void accept(
+      final HttpServerRequest request,
+      final Consumer<Optional<KafkaProducerRecord<String, CloudEvent>>> recordConsumer) {
```
Comment (slinkydeveloper) [inline data-plane/receiver/src/main/java/dev/knative/eventingkafkabroker/receiver/CloudEventRequestToRecordMapper.java]: Optional on the hot path

File: data-plane/receiver/src/main/java/dev/knative/eventingkafkabroker/receiver/CloudEventRequestToRecordMapper.java [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+
+    VertxMessageFactory.fromHttpServerRequest(request)
+        .map(Message::toEvent)
+        .onComplete(result -> {
+
+          if (result.failed()) {
+            recordConsumer.accept(Optional.empty());
```
Comment (slinkydeveloper) [inline data-plane/receiver/src/main/java/dev/knative/eventingkafkabroker/receiver/CloudEventRequestToRecordMapper.java]: In that case you need to reply to the http request too with a 400 i think

File: data-plane/receiver/src/main/java/dev/knative/eventingkafkabroker/receiver/CloudEventRequestToRecordMapper.java [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+  static final String TOPIC_DELIMITER = "-";
+
+  @Override
+  public void accept(
+      final HttpServerRequest request,
+      final Consumer<Optional<KafkaProducerRecord<String, CloudEvent>>> recordConsumer) {
+
+    VertxMessageFactory.fromHttpServerRequest(request)
+        .map(Message::toEvent)
+        .onComplete(result -> {
```
Comment (slinkydeveloper) [inline data-plane/receiver/src/main/java/dev/knative/eventingkafkabroker/receiver/CloudEventRequestToRecordMapper.java]: In this callback you should reply to the request

File: data-plane/receiver/src/main/java/dev/knative/eventingkafkabroker/receiver/RequestHandler.java [hunk 13]
```diff
@@ -0,0 +1,71 @@
+package dev.knative.eventingkafkabroker.receiver;
+
+import static io.netty.handler.codec.http.HttpResponseStatus.ACCEPTED;
```
Comment (slinkydeveloper) [inline data-plane/receiver/src/main/java/dev/knative/eventingkafkabroker/receiver/RequestHandler.java]: Use the ones from vertx `HttpStatus` class (I don't remember the specific name of the class, but i recall there is such class in vertx)

File: data-plane/receiver/src/main/java/dev/knative/eventingkafkabroker/receiver/CloudEventRequestToRecordMapper.java (hunk 12 above)
Comment (slinkydeveloper) [reply data-plane/receiver/src/main/java/dev/knative/eventingkafkabroker/receiver/CloudEventRequestToRecordMapper.java]: ah i'm dumb you did it below

File: data-plane/receiver/src/test/java/dev/knative/eventingkafkabroker/receiver/RequestHandlerTest.java [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+import io.vertx.core.Handler;
+import io.vertx.core.http.HttpServerRequest;
+import io.vertx.core.http.HttpServerResponse;
+import io.vertx.kafka.client.producer.KafkaProducer;
+import io.vertx.kafka.client.producer.KafkaProducerRecord;
+import io.vertx.kafka.client.producer.RecordMetadata;
+import java.util.Optional;
+import org.junit.jupiter.api.Test;
+
+public class RequestHandlerTest {
```
Comment (slinkydeveloper) [inline data-plane/receiver/src/test/java/dev/knative/eventingkafkabroker/receiver/RequestHandlerTest.java]: Can you do tests also performing real http requests and, on the other side, use the mocks provided by kafka client for the producer? I'm fine with mocking to check the behaviour, but i prefer to have also tests that performs the real http requests.

To create such test, you could start a single Verticle (check out https://github.com/vert-x3/vertx-junit5 for bootstrapping vertx instance in junit 5) and then perform http requests on it

**Review by slinkydeveloper — COMMENTED** [review]
Looks good so far, have you tried to compose various asynchronous operations using Promise/Future? An example: https://vertx.io/docs/guide-for-java-devs/#_wiki_verticle_initialization_phases
tldr `Promise` is the write only monad, `Future` is the read only monad. They are asynchronous `Result`/`Either`.
I'm fine with callbacks as soon as they don't _explode_ :smile:. Consider creating APIs  `Future` based (eg `RequestToRecordMapper<K, V>` could be `Function<HttpServerRequest, Future<KafkaProducerRecord<K, V>>>`), because in vertx 4 all APIs will be `Future` based

File: data-plane/receiver/src/main/java/dev/knative/eventingkafkabroker/receiver/Main.java (hunk 9 above)
Comment (pierDipi) [reply data-plane/receiver/src/main/java/dev/knative/eventingkafkabroker/receiver/Main.java]: Cool! :+1:

File: data-plane/receiver/src/main/java/dev/knative/eventingkafkabroker/receiver/RequestHandler.java (hunk 13 above)
Comment (pierDipi) [reply data-plane/receiver/src/main/java/dev/knative/eventingkafkabroker/receiver/RequestHandler.java]: Sorry, I didn't find the class you mention :(

File: .github/workflows/k8s_tests.yaml (hunk 1 above)
Comment (pierDipi) [reply .github/workflows/k8s_tests.yaml:19]: :smile:

File: data-plane/receiver/src/main/java/dev/knative/eventingkafkabroker/receiver/RequestHandler.java (hunk 13 above)
Comment (slinkydeveloper) [reply data-plane/receiver/src/main/java/dev/knative/eventingkafkabroker/receiver/RequestHandler.java]: Ok no problems, maybe it's just me that dreamed this class ahhah

File: data-plane/pom.xml [hunk 15]
```diff
@@ -84,13 +87,67 @@
...
+      <version>1.7.30</version>
+    </dependency>
+
     <dependency>
       <groupId>org.junit.jupiter</groupId>
       <artifactId>junit-jupiter</artifactId>
       <version>${junit.jupiter.version}</version>
       <scope>test</scope>
     </dependency>
+    <dependency>
```
Comment (slinkydeveloper) [inline data-plane/pom.xml:193]: You don't need this one if you use `junit-jupiter` https://github.com/junit-team/junit5-samples/blob/r5.6.2/junit5-jupiter-starter-maven/pom.xml#L18

**Comment by pierDipi** [comment]
Ready for another pass :)

File: data-plane/receiver/src/main/java/dev/knative/eventingkafkabroker/receiver/RequestToRecordMapper.java [hunk 16]
```diff
@@ -0,0 +1,22 @@
+package dev.knative.eventingkafkabroker.receiver;
+
+import io.vertx.core.http.HttpServerRequest;
+import io.vertx.kafka.client.producer.KafkaProducerRecord;
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+
+public interface RequestToRecordMapper<K, V> extends
```
Comment (slinkydeveloper) [inline data-plane/receiver/src/main/java/dev/knative/eventingkafkabroker/receiver/RequestToRecordMapper.java]: Can you convert this to future based apis?

`extends Function<HttpServerRequest, Future<KafkaProducerRecord<K, V>>>`

File: data-plane/receiver/src/test/java/dev/knative/eventingkafkabroker/receiver/integration/ReceiverVerticleTest.java [hunk 17]
```diff
@@ -0,0 +1,260 @@
...
+
+    final var httpClient = vertx.createHttpClient();
+    final var testCases = requestStream();
+    final var countDown = new CountDownLatch(testCases.size());
+    final var flags = testContext.checkpoint(testCases.size() * 2);
+
+    final var producerDriver = new ProducerDriver<String, CloudEvent>();
+    final var producer = producerDriver.producer(false);
+
+    final var handler = new RequestHandler<>(producer, new CloudEventRequestToRecordMapper());
```
Comment (slinkydeveloper) [inline data-plane/receiver/src/test/java/dev/knative/eventingkafkabroker/receiver/integration/ReceiverVerticleTest.java]: How about using the MockProducer in kafka client https://kafka.apache.org/20/javadoc/org/apache/kafka/clients/producer/MockProducer.html more than creating your custom one?

You can wrap it into vertx kafka producer with https://vertx.io/docs/apidocs/io/vertx/kafka/client/producer/KafkaProducer.html#create-io.vertx.core.Vertx-org.apache.kafka.clients.producer.Producer-

Comment (slinkydeveloper) [reply data-plane/receiver/src/test/java/dev/knative/eventingkafkabroker/receiver/integration/ReceiverVerticleTest.java]: Let's try to avoid creating mocks when we don't need these

File: data-plane/receiver/src/test/java/dev/knative/eventingkafkabroker/receiver/integration/ReceiverVerticleTest.java [hunk 18]
```diff
@@ -0,0 +1,260 @@
...
+
+    final var httpServerOptions = new HttpServerOptions();
+    httpServerOptions.setPort(PORT);
+    final var verticle = new HttpVerticle(httpServerOptions, handler);
+
+    vertx.deployVerticle(verticle, testContext.succeeding(
+        h -> testCases.parallelStream().map(tc -> tc.requestResponse).forEach(requestResponse -> {
```
Comment (slinkydeveloper) [inline data-plane/receiver/src/test/java/dev/knative/eventingkafkabroker/receiver/integration/ReceiverVerticleTest.java]: How about a parametrized test with `@ParametrizedTest`?

```
@ParameterizedTest
@MethodSource("requestStream")
public void requestHandler(TestCase case, final Vertx vertx, final VertxTestContext testContext)
```

Two things:

* Avoid creating test cases objects, It's not really junit5 pragmatic... You can use `Stream<Arguments>`: https://junit.org/junit5/docs/current/user-guide/#writing-tests-parameterized-tests-sources-MethodSource
* IMPORTANT: Put the parametrized arguments before the Vertx and VertxTestContext arguments

**Review by slinkydeveloper — COMMENTED** [review]
I would rename the package `dev.knative.eventingkafkabroker` to `dev.knative.eventing.kafka.broker`

File: data-plane/receiver/src/test/java/dev/knative/eventingkafkabroker/receiver/integration/ReceiverVerticleTest.java (hunk 17 above)
Comment (pierDipi) [reply data-plane/receiver/src/test/java/dev/knative/eventingkafkabroker/receiver/integration/ReceiverVerticleTest.java]: Using this:
```java
    final var producerMock = new MockProducer<String, CloudEvent>();
    final KafkaProducer<String, CloudEvent> producer = KafkaProducer.create(vertx, producerMock);
```
I get
```bash
io.vertx.core.VertxException: Connection was closed
```

File: data-plane/receiver/src/test/java/dev/knative/eventingkafkabroker/receiver/integration/ReceiverVerticleTest.java (hunk 18 above)
Comment (pierDipi) [reply data-plane/receiver/src/test/java/dev/knative/eventingkafkabroker/receiver/integration/ReceiverVerticleTest.java]: My goal was to send multiple messages, in this way we send only one at a time.

Comment (slinkydeveloper) [reply data-plane/receiver/src/test/java/dev/knative/eventingkafkabroker/receiver/integration/ReceiverVerticleTest.java]: you can always bootstrap the environment with setup and teardown methods (eg to start the verticle), but this way i can run a particular test with particular arguments straight from the IDE

Comment (slinkydeveloper) [reply data-plane/receiver/src/test/java/dev/knative/eventingkafkabroker/receiver/integration/ReceiverVerticleTest.java]: https://junit.org/junit5/docs/current/user-guide/#writing-tests-classes-and-methods

File: data-plane/pom.xml (hunk 15 above)
Comment (pierDipi) [reply data-plane/pom.xml:193]: Removing this produces:
```bash
[ERROR] Please refer to /app/receiver/target/surefire-reports for the individual test results.
[ERROR] Please refer to dump files (if any exist) [date].dump, [date]-jvmRun[N].dump and [date].dumpstream.
[ERROR] There was an error in the forked processorg/junit/platform/engine/EngineDiscoveryListener
[ERROR] org.apache.maven.surefire.booter.SurefireBooterForkException: There was an error in the forked processorg/junit/platform/engine/EngineDiscoveryListener
[ERROR]         at org.apache.maven.plugin.surefire.booterclient.ForkStarter.fork(ForkStarter.java:675)
[ERROR]         at org.apache.maven.plugin.surefire.booterclient.ForkStarter.run(ForkStarter.java:285)
[ERROR]         at org.apache.maven.plugin.surefire.booterclient.ForkStarter.run(ForkStarter.java:248)
[ERROR]         at org.apache.maven.plugin.surefire.AbstractSurefireMojo.executeProvider(AbstractSurefireMojo.java:1217)
[ERROR]         at org.apache.maven.plugin.surefire.AbstractSurefireMojo.executeAfterPreconditionsChecked(AbstractSurefireMojo.java:1063)
[ERROR]         at org.apache.maven.plugin.surefire.AbstractSurefireMojo.execute(AbstractSurefireMojo.java:889)
[ERROR]         at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:137)
[ERROR]         at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:210)
[ERROR]         at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:156)
[ERROR]         at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:148)
[ERROR]         at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:117)
[ERROR]         at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:81)
[ERROR]         at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:56)
[ERROR]         at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:128)
[ERROR]         at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:305)
[ERROR]         at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:192)
[ERROR]         at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:105)
[ERROR]         at org.apache.maven.cli.MavenCli.execute(MavenCli.java:957)
[ERROR]         at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:289)
[ERROR]         at org.apache.maven.cli.MavenCli.main(MavenCli.java:193)
[ERROR]         at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[ERROR]         at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[ERROR]         at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[ERROR]         at java.base/java.lang.reflect.Method.invoke(Method.java:564)
[ERROR]         at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:282)
[ERROR]         at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:225)
[ERROR]         at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:406)
[ERROR]         at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:347)
```

File: data-plane/receiver/src/test/java/dev/knative/eventingkafkabroker/receiver/integration/ReceiverVerticleTest.java (hunk 18 above)
Comment (pierDipi) [reply data-plane/receiver/src/test/java/dev/knative/eventingkafkabroker/receiver/integration/ReceiverVerticleTest.java]: I solved extracting a helper method because junit annotations don't play nicely with vertx extension.

File: data-plane/receiver/src/main/java/dev/knative/eventingkafkabroker/receiver/RequestToRecordMapper.java (hunk 16 above)
Comment (pierDipi) [reply data-plane/receiver/src/main/java/dev/knative/eventingkafkabroker/receiver/RequestToRecordMapper.java]: Done

File: data-plane/pom.xml (hunk 15 above)
Comment (slinkydeveloper) [reply data-plane/pom.xml:193]: uhm damn maven surefire plugin, that should not happen.... well add a comment explaining why you added this dep, so in future we can check out how to remove it

File: data-plane/receiver/src/main/java/dev/knative/eventing/kafka/broker/receiver/CloudEventRequestToRecordMapper.java [hunk 19]
```diff
@@ -0,0 +1,53 @@
...
+  static final int PATH_TOKEN_NUMBER = 2;
+  static final String PATH_DELIMITER = "/";
+  static final String TOPIC_DELIMITER = "-";
+
+  @Override
+  public Future<KafkaProducerRecord<String, CloudEvent>> apply(final HttpServerRequest request) {
+
+    return VertxMessageFactory.fromHttpServerRequest(request)
+        .map(Message::toEvent)
+        .map(event -> {
```
Comment (slinkydeveloper) [inline data-plane/receiver/src/main/java/dev/knative/eventing/kafka/broker/receiver/CloudEventRequestToRecordMapper.java]: nice :heart_eyes:

File: data-plane/receiver/src/main/java/dev/knative/eventing/kafka/broker/receiver/RequestHandler.java [hunk 20]
```diff
@@ -0,0 +1,66 @@
...
+    Objects.requireNonNull(requestToRecordMapper, "provide a mapper");
+
+    this.producer = producer;
+    this.requestToRecordMapper = requestToRecordMapper;
+  }
+
+  @Override
+  public void handle(final HttpServerRequest request) {
+    requestToRecordMapper.apply(request).onComplete(result -> {
+      final var record = result.result();
```
Comment (slinkydeveloper) [inline data-plane/receiver/src/main/java/dev/knative/eventing/kafka/broker/receiver/RequestHandler.java]: In case the async result failed? how do you handle it? you should do an if like:

```
if (result.succedded()) {
  // Handle success
} else {
  // Handle failure
}
```

File: data-plane/receiver/src/main/java/dev/knative/eventing/kafka/broker/receiver/RequestToRecordMapper.java [hunk 21]
```diff
@@ -0,0 +1,19 @@
+package dev.knative.eventing.kafka.broker.receiver;
+
+import io.vertx.core.Future;
+import io.vertx.core.http.HttpServerRequest;
+import io.vertx.kafka.client.producer.KafkaProducerRecord;
+import java.util.function.Function;
+
+public interface RequestToRecordMapper<K, V> extends
```
Comment (slinkydeveloper) [inline data-plane/receiver/src/main/java/dev/knative/eventing/kafka/broker/receiver/RequestToRecordMapper.java]: Annotate with `@FunctionalInterface`, it could be useful in future

File: data-plane/receiver/src/main/java/dev/knative/eventing/kafka/broker/receiver/RequestHandler.java (hunk 20 above)
Comment (pierDipi) [reply data-plane/receiver/src/main/java/dev/knative/eventing/kafka/broker/receiver/RequestHandler.java]: I think we catch it, from `AsyncResult#result()` docs:
```
The result of the operation. This will be null if the operation failed.
```

File: data-plane/receiver/src/main/java/dev/knative/eventing/kafka/broker/receiver/RequestToRecordMapper.java (hunk 21 above)
Comment (pierDipi) [reply data-plane/receiver/src/main/java/dev/knative/eventing/kafka/broker/receiver/RequestToRecordMapper.java]: Done

File: data-plane/receiver/src/main/java/dev/knative/eventing/kafka/broker/receiver/CloudEventRequestToRecordMapper.java [hunk 22]
```diff
@@ -0,0 +1,60 @@
...
+  static final String PATH_DELIMITER = "/";
+  static final String TOPIC_DELIMITER = "-";
+  public static final String TOPIC_PREFIX = "knative-";
+
+  @Override
+  public Future<KafkaProducerRecord<String, CloudEvent>> recordFromRequest(
+      final HttpServerRequest request) {
+
+    return VertxMessageFactory.fromHttpServerRequest(request)
+        // TODO is this conversion really necessary?
```
Comment (slinkydeveloper) [inline data-plane/receiver/src/main/java/dev/knative/eventing/kafka/broker/receiver/CloudEventRequestToRecordMapper.java:26]: I would say that we could avoid to convert to event but let's first complete the broker this way, then let's work to optimize it

File: data-plane/receiver/src/main/java/dev/knative/eventing/kafka/broker/receiver/RequestHandler.java [hunk 23]
```diff
@@ -0,0 +1,78 @@
...
+  }
+
+  @Override
+  public void handle(final HttpServerRequest request) {
+    requestToRecordMapper
+        .recordFromRequest(request)
+        .onSuccess(record -> send(record)
+            .onSuccess(ignore -> {
+              request.response().setStatusCode(RECORD_PRODUCED).end();
+              logger.debug("record produced - topic: {}", record.topic());
```
Comment (slinkydeveloper) [inline data-plane/receiver/src/main/java/dev/knative/eventing/kafka/broker/receiver/RequestHandler.java:60]: nice

File: data-plane/receiver/src/test/java/dev/knative/eventing/kafka/broker/receiver/integration/ReceiverVerticleTest.java [hunk 24]
```diff
@@ -117,27 +125,12 @@ private void test(
...
-    });
-  }
-
-  private void mustHave(ProducerRecord<String, CloudEvent> record) {
-    final var records = producer.history();
-    if (!records.contains(record)) {
-      fail(record + " not found\nrecords registered" + Arrays.deepToString(records.toArray()));
-    }
+    assertThat(producer.history())
+        .containsExactlyInAnyOrderElementsOf(expectedProducedRecord);
```
Comment (slinkydeveloper) [inline data-plane/receiver/src/test/java/dev/knative/eventing/kafka/broker/receiver/integration/ReceiverVerticleTest.java:127]: I love that :heart_eyes:

**Review by slinkydeveloper — APPROVED** [review]

**Comment by slinkydeveloper** [comment]
:partying_face:
